crates/dojo-core/src/world/world_contract.cairo

```diff
@@ -977,51 +870,74 @@ pub mod world {
+            let account = get_tx_info().unbox().account_contract_address;
+
+            assert(caller == account, Errors::NOT_A_DIRECT_CALL);
+        }
+
+        /// Panic if the caller is NOT an owner of the resource.
         ///
         /// # Arguments
-        ///
-        /// * `resource` - The selector of the resource being verified.
-        ///
-        /// # Returns
-        ///
-        /// * `bool` - True if the calling account is the owner of the resource or the owner of the
-        /// world,
-        ///            false otherwise.
+        ///   * `resource_id` - the id of the resource.
         #[inline(always)]
-        fn is_account_owner(self: @ContractState, resource: felt252) -> bool {
-            IWorld::is_owner(self, resource, self.get_account_address())
-                || self.is_account_world_owner()
+        fn must_be_resource_owner(self: @ContractState, resource_id: felt252) {
+            assert(
+                self.is_caller_world_owner() || self.is_owner(resource_id, get_caller_address()),
+                Errors::NOT_OWNER
+            );
         }
 
-        /// Verifies if the calling account has write access to the resource.
+        /// Panic if the caller is NOT allowed to write in the namespace.
         ///
         /// # Arguments
-        ///
-        /// * `resource` - The selector of the resource being verified.
-        ///
-        /// # Returns
-        ///
-        /// * `bool` - True if the calling account has write access to the resource,
-        ///            false otherwise.
+        ///   * `namespace_hash` - the hash of the namespace.
         #[inline(always)]
-        fn is_account_writer(self: @ContractState, resource: felt252) -> bool {
-            IWorld::is_writer(self, resource, self.get_account_address())
+        fn must_have_namespace_write_access(self: @ContractState, namespace_hash: felt252) {
+            let caller = get_caller_address();
+            assert(
+                self.is_caller_world_owner()
+                    || self.is_owner(namespace_hash, caller)
+                    || self.is_writer(namespace_hash, caller),
+                Errors::NO_NAMESPACE_WRITE_ACCESS
+            );
         }
 
-        /// Verifies if the calling account is the world owner.
+        /// Panic if the caller is NOT allowed to write in the model.
         ///
-        /// # Returns
-        ///
-        /// * `bool` - True if the calling account is the world owner, false otherwise.
+        /// # Arguments
+        ///   * `model_selector` - the selector of the model.
         #[inline(always)]
-        fn is_account_world_owner(self: @ContractState) -> bool {
-            IWorld::is_owner(self, WORLD, self.get_account_address())
+        fn must_have_model_write_access(self: @ContractState, model_selector: felt252) {
+            let caller = get_caller_address();
+
+            // must have owner or writer role on the model ...
```

Comment:
Maybe checking first for namespace could optimize most of the calls as the rights are now mostly given at namespace level first.

This should save some steps as use extensively in Dojo.